src/Tmds.MDns/MdnsSocket.cs

```diff
@@ -0,0 +1,142 @@
+//Copyright (C) 2013  Tom Deseyn
+
+//This library is free software; you can redistribute it and/or
+//modify it under the terms of the GNU Lesser General Public
+//License as published by the Free Software Foundation; either
+//version 2.1 of the License, or (at your option) any later version.
+
+//This library is distributed in the hope that it will be useful,
+//but WITHOUT ANY WARRANTY; without even the implied warranty of
+//MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+//Lesser General Public License for more details.
+
+//You should have received a copy of the GNU Lesser General Public
+//License along with this library; if not, write to the Free Software
+//Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.NetworkInformation;
+using System.Net.Sockets;
+using System.Threading;
+using System.Runtime.InteropServices;
+
+namespace Tmds.MDns
+{
+    class MdnsSocket : IDisposable
+    {
+        public static readonly IPEndPoint IPv4EndPoint = new IPEndPoint(IPAddress.Parse("224.0.0.251"), 5353);
+        public static readonly IPEndPoint IPv6EndPoint = new IPEndPoint(IPAddress.Parse("ff02::fb"), 5353);
+
+        public static MdnsSocket CreateSocketIPv4(int index)
+        {
+            var socket = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
+            socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, true);
+
+            socket.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.MulticastInterface, IPAddress.HostToNetworkOrder(index));
+
+            socket.Bind(new IPEndPoint(IPAddress.Any, IPv4EndPoint.Port));
+            var tempEndpoint = new IPEndPoint(IPAddress.Any, IPv4EndPoint.Port);
+
+            IPAddress ip = IPv4EndPoint.Address;
+            socket.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.AddMembership, new MulticastOption(ip, index));
+            socket.SetSocketOption(SocketOptionLevel.IP, SocketOptionName.MulticastTimeToLive, 1);
+            var info = new MdnsSocket(socket, IPv4EndPoint);
+            return info;
+        }
+
+        public static MdnsSocket CreateSocketIPv6(int index)
+        {
+            var socket = new Socket(AddressFamily.InterNetworkV6, SocketType.Dgram, ProtocolType.Udp);
+            socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, true);
+
+            // The MulticastInterface fails for IPv6, see: https://github.com/dotnet/runtime/issues/24255
```

Comment:
This is an old issue about MulticastInterface not working for IPv6 on Linux.

By now, we should be able to:
```
socket.SetSocketOption(SocketOptionLevel.IPv6, SocketOptionName.MulticastInterface, interfaceIndex).
```
and this should work on Linux and Windows afaik.

Reply:
Yes, this works now. I did try this before but couldn't get it to work. Guess there were some other issues at that time. It works now on both Linux and Windows.